### ISSUE TITLE
Bridge speed and flow ratio is used on incorrect solid infill layers.

### ISSUE THREAD

**Issue by TheBoojah** [alexrj/Slic3r#4172]
### Version
1.2.9 and latest dev (oct 23)

### Operating system type + version
Win 10 latest patch

### Behavior
Bridge speed and flow ratio is used on incorrect solid infill layers.
Bridge should be use on the lowest top-surface solid layer, i.e. the first layer on top of the infill.
What I am getting is all solid layers except the very top and bottom ones use bridge speed and flow. So for 3 bottom 3 top, the first is ok, then 2 bridge layers - infill - 2 bridge and the final top layer.

To illustrate I take all the default settings, except changing bridge flow ratio to 0.1 to make it more visible:
![image](https://user-images.githubusercontent.com/33135012/32073850-b5f4836e-ba97-11e7-8562-183584033136.png)
The solid infill lines are very thin, indicating that bridge settings were used. 
This is just a 20x20x20 cube, and I believe there is no other settings changed in Slic3r - everything is default.

If this is not easily reproducible please let me know what additional info I can provide.
Thanks

**Comment by lordofhyphens** [comment]
Please zip that model and an export config (Ctrl-E) and attach it to the issue (drag and drop)

**Comment by TheBoojah** [comment]
I updated the issue with the requested files.

**Comment by lordofhyphens** [comment]
As far as I am aware, the first "solid" infill over sparse infill is bridge settings by design (as that's really what it is, a bridge).

**Comment by lordofhyphens** [comment]
It appears that all internal solid infill is being flagged as bridges, beyond the first layer above sparse infill.

It's definitely wrong for bottom layers that aren't above a void.

**Comment by lordofhyphens** [comment]
Other notes: 
At 100% infill bridge infill is not used at all, so there's a special case going on here.

If you wanted to have a workaround for a simple case you can set the main mesh to 100% infill and make a modifier mesh that is scaled down slightly and set to the real infill density. Slic3r does not seem to pay attention to infill differences for modifier meshes for the purposes of internal bridge infill.

**Comment by lordofhyphens** [comment]
I am pretty sure the error is in the surface type detection. Top and bottom surfaces are easy to find, so those are classified correctly. 

The associated code should be in Layer::detect_surfaces_type().

It's actually applied around line 205 of LayerRegionFill.cpp where the flow is calculated.

**Comment by TheBoojah** [comment]
Thanks for looking into this bug!

The way I detected this is playing with the bridge settings, at 0.8 bridge flow ratio the thinner lines starts to get really noticeable in the preview. The real killer is however when it does 10mm/s on the second solid layer of a larger part. So I can se why nobody has reported this to far, as when you have default or "light" bridge settings you never notice a difference. On challenging parts with some more extreme bridge settings can really mess things up, a high bridge fan setting on a large second solid layer is not so great for bed adhesion for example.

Thanks for looking for a workaround. It won't work for my current model so another way is just to use close to default bridge settings and hope the bridges come out ok anyway.
 
I tested Slic3r 1.1.2 which worked correctly, only on the first solid layer over sparse infill is bridge settings used.

**Comment by lordofhyphens** [comment]
Yeah, this is a regression from when the logic was ported.

**Comment by wirlybird** [comment]
Hey guys, came across your posts searching for a solution to a similar issue.  This pic is the first layer over a support.  It is being treated as a bridge and making the extrusions too thin so they mainly do not attach to anything and when you remove the supports you have a stringy mess.
Increasing the flow percent just pushes the extrusions further apart so the gaps stay the same and it doesn't solve anything.
Now, I can understand why it may think it needs bridging here but it just makes a mess.

If I run the same part in S3D it builds a solid layer over the support and it is all good.

Would love to find a solution!  Even just being able to completely disable bridging would be good.

![1](https://user-images.githubusercontent.com/25409320/34323843-32786918-e820-11e7-8250-ef73e412f6a5.png)

![20171223_104333](https://user-images.githubusercontent.com/25409320/34323857-729df4f4-e820-11e7-91bb-e1e1bfbe1dbc.jpg)

**Comment by lordofhyphens** [comment]
It *is* a bridge. A bridge is any extrusion that doesn't have something
underneath it. In these case, things like layer height and width are
meaningless.

**Comment by wirlybird** [comment]
Thanks, I understand it being a "bridge".  However, in a circumstance like this it creates an interface layer like in my example.
I looked at the same part in S3D and it also treats it as a bridge but the extrusions are tight together so it makes a good solid layer.  No saying one slicer is better than another!
Is there anything we can do to improve this first layer on top of supports?  Can we force it to a standard infill and perimeter layer?  Is there a way to even  just get standard extrusion widths?
Slic3r does a great job on this particular part of mine except for this one issue and is why I am using it.

I appreciate any ideas and happy holidays!!

**Comment by lordofhyphens** [comment]
Width here is your nozzle diameter. No more, no less. You could tune the
placement to be more tightly packed but then we run the extra risk of
destroying the bridge with nozzle drag.

For this specific model, you could add a modifier mesh to set infill to
100% below the problematic area.

**Comment by wirlybird** [comment]
I understand and thanks for the info.
I'll see if I can figure out how to do a modifier mesh!  Sounds like something I need to know!

**Comment by wirlybird** [comment]
I found a tutorial to do the modifier mesh and I can get it to change the "infill" but I still can't seem to get it to affect the first layer over the support!!
Interesting all the same.

**Comment by lordofhyphens** [comment]
Right, first layer over the support. I was thinking of something else.

You could set the interface spacing to 0, which might help a little. Or see
if you can change the bridge flow rate % at your modifier mesh (I do not
remember if this is possible).

**Comment by wirlybird** [comment]
Thanks.  I tried messing with the bridge flow rate but it doesn't seem to make a difference.
I'll see about the interface spacing.

**Comment by bubnikv** [comment]
This may be of interest to you.
https://github.com/prusa3d/Slic3r/issues/102
https://github.com/prusa3d/Slic3r/issues/612

Namely, there is a workaround to force Slic3r to squish the bridging
threads against the support interface layer.

**Comment by wirlybird** [comment]
Thanks for the info.  Interesting reading.
